### PR TITLE
Do SSL Client auth in webview using the certificate configured for the

### PR DESCRIPTION
**Pull request by rcatolino**
This commit uses the certificate configured for the openhab server in webview widgets as well.

Ideally the user should be able to choose another certificate if needed, but I this fixes the case where the webview URL is behind the same reverse proxy as the openhab server and the certificate needed is the same.

Signed-off-by: rca <raphael.catolino@gmail.com>

### CODE REVIEW

File: mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt [hunk 1]
```diff
@@ -62,6 +69,41 @@ open class ConnectionWebViewClient(
...
+        if (prefs.isDemoModeEnabled()) {
+            request.cancel()
+        }
+
+        val alias = prefs.getString(Constants.PREFERENCE_SSL_CLIENT_CERT, null)
+        Log.d(TAG, "Using alias " + alias)
+        if (alias == null) {
+            request.cancel()
+        } else {
+            ForwardRequestTask(request, view).execute(alias)
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]: This should be implemented using a coroutine via `launch` block instead.

File: mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt [hunk 2]
```diff
@@ -62,6 +70,34 @@ open class ConnectionWebViewClient(
...
+        Log.d(TAG, "SSL Cient Cert required")
+        val prefs = view.context.getPrefs()
+        if (prefs.isDemoModeEnabled()) {
+            request.cancel()
+        }
+
+        val alias = prefs.getString(Constants.PREFERENCE_SSL_CLIENT_CERT, null)
+        Log.d(TAG, "Using alias " + alias)
+        if (alias == null) {
+            request.cancel()
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]: You can return here and remove the `else`.

File: mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt [hunk 3]
```diff
@@ -62,6 +71,36 @@ open class ConnectionWebViewClient(
...
+    override fun onReceivedClientCertRequest(view: WebView, request: ClientCertRequest) {
+        Log.d(TAG, "SSL Cient Cert required")
+        val prefs = view.context.getPrefs()
+        if (prefs.isDemoModeEnabled()) {
+            request.cancel()
+            return
+        }
+
+        val alias = prefs.getString(Constants.PREFERENCE_SSL_CLIENT_CERT, null)
+        Log.d(TAG, "Using alias " + alias)
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]:
```suggestion
        Log.d(TAG, "Using alias $alias")
```

File: mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt [hunk 4]
```diff
@@ -62,6 +71,36 @@ open class ConnectionWebViewClient(
...
+            return
+        }
+
+        GlobalScope.launch(Dispatchers.IO) {
+            try {
+                val chain = KeyChain.getCertificateChain(view.context, alias)
+                val privateKey = KeyChain.getPrivateKey(view.context, alias)
+                request.proceed(privateKey, chain)
+            } catch (e: KeyChainException) {
+                Log.d(TAG, "KeyChain Exception", e)
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]:
```suggestion
                Log.d(TAG, "Error getting chain or private key", e)
```

The exception itself is also passed the log line, so the string doesn't need to contain the exception name.

File: mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt [hunk 5]
```diff
@@ -62,6 +71,36 @@ open class ConnectionWebViewClient(
...
+                request.ignore()
+            } catch (e: InterruptedException) {
+                Log.d(TAG, "Interrupted Exception", e)
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]: Same as above.

File: mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt [hunk 6]
```diff
@@ -62,6 +71,36 @@ open class ConnectionWebViewClient(
         }
     }
 
+    override fun onReceivedClientCertRequest(view: WebView, request: ClientCertRequest) {
+        Log.d(TAG, "SSL Cient Cert required")
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]:
```suggestion
        Log.d(TAG, "SSL Client Cert required")
```

File: mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt [hunk 7]
```diff
@@ -62,6 +71,36 @@ open class ConnectionWebViewClient(
...
+        if (alias == null) {
+            request.cancel()
+            return
+        }
+
+        GlobalScope.launch(Dispatchers.IO) {
+            try {
+                val chain = KeyChain.getCertificateChain(view.context, alias)
+                val privateKey = KeyChain.getPrivateKey(view.context, alias)
+                request.proceed(privateKey, chain)
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]: I wonder whether it's OK to do the callback on a different thread, or whether we should do it in the main thread.

Comment (rcatolino) [reply mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]: FWIW I haven't had any issues using it so far so it doesn't seem to cause any problems.
I'm actually wondering the opposite : the whole class is only used for networking stuff, shouldn't all that be happening on the IO thread ?
(Note that I don't know the first thing about kotlin, and I haven't worked on Android for ages, so I don't really know what I'm talking about.)

Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/ConnectionWebViewClient.kt]: I looked into the docs for this, and the current behavior seems fine, so NVM.

**Pull request merged**